core/lib/spree/migration_helpers.rb

```diff
@@ -0,0 +1,64 @@
+module Spree
+  module MigrationHelpers
+    class SafeIndexChange
+      attr_reader :migration, :table, :columns, :options
+
+      def initialize(migration, table, columns, options = {})
+        @migration = migration
+        @table = table
+        @columns = columns
+        @options = options
+      end
+
+      def remove_index
+        migration.remove_index(table, columns) if index_present?
+      end
+
+      def add_index
+        if columns.is_a?(Array)
+          add_multi_column_index
+        else
+          add_single_column_index
+        end
+      end
+
+      private
+
+      def index_present?
+        migration.index_exists?(table, columns, options)
+      end
+
+      def index_absent?
+        !index_present?
+      end
+
+      def add_multi_column_index
+        if all_columns_exist? && index_absent?
+          migration.add_index(table, columns, options)
+        end
+      end
+
+      def all_columns_exist?
+        columns.all? { |column| migration.column_exists?(table, column) }
```

Comment:
My initial reaction to writing a class for this was just like yours. It does have two benefits:
- The public methods read very nicely.
- Single column index additions get the options parameter passed along to `column_exists?`

If you feel this is too heavy weight for those benefits, I'll go back to the block-inside-conditional style like you've suggested.